truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/source/SourceSection.java

```diff
@@ -76,6 +83,30 @@
         this.startColumn = startColumn;
         this.charIndex = charIndex;
         this.charLength = charLength;
+        this.tags = tags;
+        assert tagsAreNonNullAndInterned(tags) : "All tags set for a source section must be interned and non-null.";
+    }
+
+    @SuppressFBWarnings("ES_COMPARING_STRINGS_WITH_EQ")
+    private static boolean tagsAreNonNullAndInterned(String[] tags) {
+        for (int i = 0; i < tags.length; i++) {
+            if (tags[i] == null) {
+                return false;
+            }
+            if (tags[i].intern() != tags[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Returns a set of tags that are associated with this source section as array. The returned
+     * array must be considered read-only. All elements of the array were asserted to be non-null
+     * and interned for faster comparison.
+     */
+    public String[] getTags() {
```

Comment:
If our goal is to turn **Truffle** from an academic research into a product, we need to stop the *"try & deprecate later"* attitude. API design is about balancing freedom - freedom of API designer and freedom of API users. In this case we (as API designers) are opening up twice:

- by allowing users to use any string as a tag
- by allowing any user to read the tags associated with any SourceSection

As I mentioned at initial review, I believe using strings (e.g. case 1) is suboptimal. I am (as a minority) advising not to do it. 

I believe giving everyone a chance to read the tags (e.g. case 2) is even bigger mistake and will paint us into a corner. I am strongly advising not to do it.

Btw.  I may even share the passion to allow access to all data provided in the constructor, but I don't think we have use-cases for that at the moment. On contrary - we are unsure whether the data passed in should even be there...

Anyway I am not going to block anything. If your desire is to give up on your freedom in these two areas, it is your decision. I just want you to understand the consequences - e.g. once you give this freedom to API users, I promise to fight on their behalf to make sure their just gained freedom is *never taken away* from them.